### PR TITLE
spark: publish ProcessingEngineRunFacet

### PR DESCRIPTION
**Pull request by d-m-h**
### Problem

The Spark integration publishes a custom `SparkVersionFacet` that isn't defined in the spec, yet there is a facet that captures this information (and a bit more) called `ProcessingEngineRunFacet`.

Closes: [2086](https://github.com/OpenLineage/OpenLineage/issues/2086)

### Solution

Previously, the Spark integration published a custom facet named 'SparkVersion'. As it is custom, it isn't defined in the OpenLineage spec. However, the OpenLineage spec defines a ProcessingEngineRunFacet that is meant to capture details about the things that runs a job.

This change introduces this in the form of the
'SparkProcessingEngineRunFacetBuilder' and
'SparkProcessingEngineRunFacetBuilderDelegate'.

As the names suggest, these classes are meant to create and populate ProcessingEngineRunFacet.

The reason for the existence of the delegate is because there are two code paths that interact with the run facets, namely the code path within the RddExecutionContext and the other in the SparkSqlExecutionContext, albeit via a very roundabout way.

The delegate is the object that actually constructs the facet, whilst the builder provides an adapter that uses the CustomFacetBuilder interface.

Yes, its a hacky way of doing it and may need to be changed in the future. For now though, its good enough.

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [x] You've updated any relevant documentation (_if relevant_)
- [x] Your comment includes a one-liner for the changelog about the specific purpose of the change (_if necessary_)
- [x] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/SparkProcessingEngineRunFacetBuilderDelegate.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  public ProcessingEngineRunFacet buildFacet() {
+    // TODO: At some stage, we probably need to create a canonical dictionary for the names of the
+    //  processing engines, or define a convention - i.e., names of processing engines should be
+    //  snake_case
+    return ol.newProcessingEngineRunFacetBuilder()
+        .name("spark")
+        .version(sparkContext.version())
+        .openlineageAdapterVersion(this.getClass().getPackage().getImplementationVersion())
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/SparkProcessingEngineRunFacetBuilderDelegate.java:40]: We do have a spark version field within `OpenLineageContext` object

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,10 @@
 # Changelog
 
 ## [Unreleased](https://github.com/OpenLineage/OpenLineage/compare/1.1.0...HEAD)
+### Feature
+* **Spark: Publish the `ProcessingEngineRunFacet` as part of the normal operation of the `OpenLineageSparkEventListener`.** [`#2089`](https://github.com/OpenLineage/OpenLineage/pull/2089) [@d-m-h](https://github.com/d-m-h)
+  *Publishes the spec-defined `ProcessEngineRunFacet` alongside the custom `SparkVersionFacet`.*
```
Comment (pawel-big-lebowski) [inline CHANGELOG.md:6]: I think we should mark `SparkVersionFacet` deprecated and mention OL version where we will get rid of it (like two versions ahead).

File: integration/spark/shared/src/test/java/io/openlineage/spark/agent/facets/builder/SparkProcessingEngineFacetBuilderTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.spark.scheduler.SparkListenerStageSubmitted;
+import org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionEnd;
+import org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionStart;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import scala.collection.Seq$;
+
+class SparkProcessingEngineFacetBuilderTest {
+  private static SparkContext sparkContext;
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/test/java/io/openlineage/spark/agent/facets/builder/SparkProcessingEngineFacetBuilderTest.java:31]: Please consider using a mock intstead.

**Review by pawel-big-lebowski — COMMENTED** [review]
This is the first-rate PR. Thank you for this 👍 🥇 

I put some comments. Let me know what do you think about it. 

Please don't worry about failing integration tests. 
It's normal for ci build from external forks. 
We will help you with this and sorry for inconvenience.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/SparkProcessingEngineRunFacetBuilderDelegate.java (hunk 1 above)
Comment (d-m-h) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/SparkProcessingEngineRunFacetBuilderDelegate.java:40]: Aye, I initially used that. However, we don't have an `OpenLineageContext` object available in the `RddExecutionContext` code path, and refactoring the `RddExecutionContext` is more work than I am comfortable taking on right now, given that I am new to the project.

Chances are, we may need to make changes to the `OpenLineageContext` definition, due to its requirement of having a `QueryExecution` object embedded within it, which as far as I know, is only available when you're dealing with Spark's DataFrame API - but I stand to be corrected on this.

File: integration/spark/shared/src/test/java/io/openlineage/spark/agent/facets/builder/SparkProcessingEngineFacetBuilderTest.java (hunk 3 above)
Comment (d-m-h) [reply integration/spark/shared/src/test/java/io/openlineage/spark/agent/facets/builder/SparkProcessingEngineFacetBuilderTest.java:30]: Done.

File: CHANGELOG.md (hunk 2 above)
Comment (d-m-h) [reply CHANGELOG.md]: Done. Updated the change long as well.

**Review by pawel-big-lebowski — APPROVED** [review]
Looks good to me.

**Pull request merged**